tests/ctest/test_opts.c

```diff
@@ -0,0 +1,779 @@
+	succeed_if (checkError (errorKey, xstr (ELEKTRA_ERROR_OPTS_UNKNOWN_OPTION), "Unknown short option: -s"),
+		    "short help with value (with arg, combined) should have failed");
+	clearValues (ks);
+
+	RUN_TEST_ERROR (ks, errorKey, ARGS ("--help=long"), NO_ENVP);
+	succeed_if (checkError (errorKey, xstr (ELEKTRA_ERROR_OPTS_ILLEGAL_USE), "This option cannot have an argument: --help"),
+		    "long help with value (with arg, combined) should have failed");
+	clearValues (ks);
+
+	ksDel (ks);
+}
+
+static void test_stop (void)
+{
+	KeySet * ks = ksNew (1, keyWithOpt (SPEC_BASE_KEY "/apple", 'a', "apple", "APPLE"),
+			     keyNew (SPEC_BASE_KEY "/rest/#", KEY_META, "args", "remaining", KEY_END), KS_END);
+
+	RUN_TEST (ks, ARGS ("--", "-a", "short"), NO_ENVP);
+	succeed_if (checkValue (ks, PROC_BASE_KEY "/apple", NULL), "should have stopped");
+	succeed_if (checkValue (ks, PROC_BASE_KEY "/rest", "#1"), "rest has wrong count");
+	succeed_if (checkValue (ks, PROC_BASE_KEY "/rest/#0", "-a"), "rest has wrong value (#0)");
+	succeed_if (checkValue (ks, PROC_BASE_KEY "/rest/#1", "short"), "rest has wrong value (#1)");
+	clearValues (ks);
+
+	RUN_TEST (ks, ARGS ("-ashort", "--", "-a", "short2"), NO_ENVP);
+	succeed_if (checkValue (ks, PROC_BASE_KEY "/apple", "short"), "should have stopped after short option");
+	succeed_if (checkValue (ks, PROC_BASE_KEY "/rest", "#1"), "rest has wrong count");
+	succeed_if (checkValue (ks, PROC_BASE_KEY "/rest/#0", "-a"), "rest has wrong value (#0)");
+	succeed_if (checkValue (ks, PROC_BASE_KEY "/rest/#1", "short2"), "rest has wrong value (#1)");
+	clearValues (ks);
+
+	RUN_TEST (ks, ARGS ("--apple", "long", "--", "-a", "short2"), NO_ENVP);
+	succeed_if (checkValue (ks, PROC_BASE_KEY "/apple", "long"), "hould have stopped after long option");
+	succeed_if (checkValue (ks, PROC_BASE_KEY "/rest", "#1"), "rest has wrong count");
+	succeed_if (checkValue (ks, PROC_BASE_KEY "/rest/#0", "-a"), "rest has wrong value (#0)");
+	succeed_if (checkValue (ks, PROC_BASE_KEY "/rest/#1", "short2"), "rest has wrong value (#1)");
+	clearValues (ks);
+
+	RUN_TEST (ks, ARGS ("--"), ENVP ("APPLE=env"));
+	succeed_if (checkValue (ks, PROC_BASE_KEY "/apple", "env"), "env-var failed (stopped options)");
+	succeed_if (checkValue (ks, PROC_BASE_KEY "/rest", "#"), "rest has wrong count");
+	clearValues (ks);
+
+
+	Key * errorKey = keyNew ("spec/tests/opts", KEY_META, "posixly", "1", KEY_END);
+	if (elektraGetOpts (ks, ARGS ("-ashort", "other", "-a", "short2"), NO_ENVP, errorKey) != 0)
+	{
+		yield_error ("error found");
+		output_error (errorKey);
+	}
+	keyDel (errorKey);
+	succeed_if (checkValue (ks, PROC_BASE_KEY "/apple", "short"), "should have stopped after short option");
+	succeed_if (checkValue (ks, PROC_BASE_KEY "/rest", "#2"), "rest has wrong count");
+	succeed_if (checkValue (ks, PROC_BASE_KEY "/rest/#0", "other"), "rest has wrong value (#0)");
+	succeed_if (checkValue (ks, PROC_BASE_KEY "/rest/#1", "-a"), "rest has wrong value (#1)");
```

Comment:
Shouldn't that be an error? (Specifying -a again without argument?)

Reply:
No, this test uses the "posixly" option (same as `POSIXLY_CORRECT` for getopt_long(3)). Therefore option processing should stop before "other".

Reply:
If we really want to support `POSIXLY_CORRECT`, then we maybe should also support it as configuration setting and not let it hard-code within applications.

The question is if we want `POSIXLY_CORRECT` only as environment variable or as Elektra configuration setting (also in config files next to env var).

Reply:
I don't think `POSIXLY_CORRECT` should be user configurable. As the default suggests, not using `POSIXLY_CORRECT` is the way to go in most cases. The non-posix mode is more flexible for the user.

But if you use an interface like `git` uses, you would rely on `POSIXLY_CORRECT`. You program would break, if the user disables it, because there would be no way to know whether an option was given to the main program itself (i.e. before the subcommand in argv) or to the subcommand (i.e. after the subcommand in argv).

Reply:
You mean relying on that `POSIXLY_CORRECT`  is **not** set?

`git -p diff` and `git diff -p` do the same thing, what do you mean with "options passed to subcommand vs. command"? With `POSIXLY_CORRECT` (if git would implement it) users could always do `git -p diff`. It would be strange but it should work?

The whole idea of `POSIXLY_CORRECT` is to have a hack for situations where "POSIX and common sense disagree". I do not know if anyone actually uses this option.

Reply:
> git -p diff and git diff -p do the same thing

That is actually a coincidence. The `-p` before `diff` is short for `--paginate` and pipes all output through `less`. The `-p` after `diff` is short for `--patch`, which is the default option for `diff` and can therefore be omitted.

Try the same with `-u` which is an alternative for `-p`. `git -p diff` will result in "unknown option". As far as I can tell the reasoning behind this behaviour is to allow general `git` options and subcommand options to use the same names.